### PR TITLE
1 database setup

### LINKED ISSUE #1: Database Setup
name: Feature request
about: User Story
title: ''
labels: enhancement
assignees: ''

---

**Put the User Story Here.**

### PR DESCRIPTION
**Pull request by Daniel-Gallagher92**
https://github.com/z-fitch/little-shop-7/blob/main/doc/db_setup.md

### CODE REVIEW

File: lib/tasks/csv_load.rake [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    #reset pk sequence 
+    ActiveRecord::Base.connection.reset_pk_sequence!('customers')
+    puts "done"
+  end
+  
+  desc "load invoice_items"
+  task invoice_items: :environment do 
+    #access CSV file
+    file = "db/data/invoice_items.csv"
+    Customer.destroy_all
```
Comment (z-fitch) [inline lib/tasks/csv_load.rake]: We gotta change these

**Review by ethanrossblack — APPROVED** [review]
We did this together on Zoom, looks good

**Review by ethanrossblack — CHANGES_REQUESTED** [review]
I rescind my initial approval after @z-fitch pointed out some good errors to fix

**Review by z-fitch — CHANGES_REQUESTED** [review]
rake file:Change Customer to the correct Object (merchant, etc) in the .create statements, and .destroy

File: lib/tasks/csv_load.rake [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+end
+
+
+
+# desc "leafblower"
+# task csv_load: [:merchants, :customers, :items, :invoices, :invoice_items, :transactions] do 
+
+
+# end
```
Comment (ethanrossblack) [inline lib/tasks/csv_load.rake]:
```suggestion
```

File: spec/models/customer_spec.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+require 'rails_helper'
+
+RSpec.describe Customer, type: :model do 
+  it "creates a customer" do 
+    
+  end
```
Comment (ethanrossblack) [inline spec/models/customer_spec.rb]:
```suggestion
```

Don't need to test "creation" right now

**Review by ethanrossblack — CHANGES_REQUESTED** [review]

File: lib/tasks/csv_load.rake [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    #reset pk sequence 
+    ActiveRecord::Base.connection.reset_pk_sequence!('transactions')
+    puts "done"
+  end
+
+  desc "load all"
+  # task :all do 
+  #   Rake::Task
+    task all: [:merchants, :customers, :items, :invoices, :invoice_items, :transactions]
+  end
```
Comment (ethanrossblack) [inline lib/tasks/csv_load.rake]:
```suggestion
  task all: [:merchants, :customers, :items, :invoices, :invoice_items, :transactions] do
  end
```

remove the comments, add a "do" to the end of the `task all:` line

File: lib/tasks/csv_load.rake [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    file = "db/data/customers.csv"
+    Customer.destroy_all
+    #parse CSV file
+    CSV.foreach(file, headers: true) do |row|
+      #convert data into models for db
+      Customer.create(row.to_h)
+    end
+    #reset pk sequence 
+    ActiveRecord::Base.connection.reset_pk_sequence!('customers')
+    puts "done"
```
Comment (ethanrossblack) [inline lib/tasks/csv_load.rake]:
```suggestion
    puts "Customers loaded from CSV"
```

File: lib/tasks/csv_load.rake [hunk 6]
```diff
@@ -0,0 +1,101 @@
+require 'csv'
+
+namespace :csv_load do 
+
+  desc "load customers"
```
Comment (ethanrossblack) [inline lib/tasks/csv_load.rake]:
```suggestion
  desc "Load customers from CSV data"
```

File: lib/tasks/csv_load.rake [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+      Item.destroy_all
+      #parse CSV file
+      CSV.foreach(file, headers: true) do |row|
+        #convert data into models for db
+        Item.create(row.to_h)
+      end
+      #reset pk sequence 
+      ActiveRecord::Base.connection.reset_pk_sequence!('items')
+      puts "done"
+  end
```
Comment (ethanrossblack) [inline lib/tasks/csv_load.rake]:
```suggestion
  desc "Load items from CSV data"
  task items: :environment do 
    file = "db/data/items.csv"
    Item.destroy_all
    CSV.foreach(file, headers: true) do |row|
      Item.create(row)
    end
    ActiveRecord::Base.connection.reset_pk_sequence!('items')
    puts "Items loaded from CSV"
  end
```

File: lib/tasks/csv_load.rake [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+  desc "load customers"
+  task customers: :environment do 
+    #access CSV file
+    file = "db/data/customers.csv"
+    Customer.destroy_all
+    #parse CSV file
+    CSV.foreach(file, headers: true) do |row|
+      #convert data into models for db
+      Customer.create(row.to_h)
```
Comment (ethanrossblack) [inline lib/tasks/csv_load.rake]:
```suggestion
      Customer.create(row)
```

I tested and we don't need `.to_h`. It still creates customers if you only use `row`.

File: lib/tasks/csv_load.rake [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    InvoiceItem.destroy_all
+    #parse CSV file
+    CSV.foreach(file, headers: true) do |row|
+      #convert data into models for db
+      InvoiceItem.create(row.to_h)
+    end
+    #reset pk sequence 
+    ActiveRecord::Base.connection.reset_pk_sequence!('invoice_items')
+    puts "done"
+  end
```
Comment (ethanrossblack) [inline lib/tasks/csv_load.rake]:
```suggestion
  desc "Load invoice_items from CSV data"
  task invoice_items: :environment do 
    file = "db/data/invoice_items.csv"
    InvoiceItem.destroy_all
    CSV.foreach(file, headers: true) do |row|
      InvoiceItem.create(row)
    end
    ActiveRecord::Base.connection.reset_pk_sequence!('invoice_items')
    puts "Invoice_Items loaded from CSV"
  end
```

File: lib/tasks/csv_load.rake [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    Invoice.destroy_all
+    #parse CSV file
+    CSV.foreach(file, headers: true) do |row|
+      #convert data into models for db
+      Invoice.create(row.to_h)
+    end
+    #reset pk sequence 
+    ActiveRecord::Base.connection.reset_pk_sequence!('invoices')
+    puts "done"
+  end
```
Comment (ethanrossblack) [inline lib/tasks/csv_load.rake]:
```suggestion
  desc "Load invoices from CSV data"
  task invoices: :environment do 
    file = "db/data/invoices.csv"
    Invoice.destroy_all
    CSV.foreach(file, headers: true) do |row|
      Invoice.create(row)
    end
    ActiveRecord::Base.connection.reset_pk_sequence!('invoices')
    puts "Invoices loaded from CSV"
  end
```

File: lib/tasks/csv_load.rake [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+      Merchant.destroy_all
+      #parse CSV file
+      CSV.foreach(file, headers: true) do |row|
+        #convert data into models for db
+        Merchant.create(row.to_h)
+      end
+      #reset pk sequence 
+      ActiveRecord::Base.connection.reset_pk_sequence!('merchants')
+      puts "done"
+  end
```
Comment (ethanrossblack) [inline lib/tasks/csv_load.rake]:
```suggestion
  desc "Load merchants from CSV data"
  task merchants: :environment do 
    file = "db/data/merchants.csv"
    Merchant.destroy_all
    CSV.foreach(file, headers: true) do |row|
      Merchant.create(row)
    end
    ActiveRecord::Base.connection.reset_pk_sequence!('merchants')
    puts "Merchants loaded from CSV"
  end
```

File: lib/tasks/csv_load.rake [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    Transaction.destroy_all
+    #parse CSV file
+    CSV.foreach(file, headers: true) do |row|
+      #convert data into models for db
+      Transaction.create(row.to_h)
+    end
+    #reset pk sequence 
+    ActiveRecord::Base.connection.reset_pk_sequence!('transactions')
+    puts "done"
+  end
```
Comment (ethanrossblack) [inline lib/tasks/csv_load.rake]:
```suggestion
  desc "Load transactions from CSV data"
  task transactions: :environment do 
    file = "db/data/transactions.csv"
    Transaction.destroy_all
    CSV.foreach(file, headers: true) do |row|
      Transaction.create(row.to_h)
    end
    ActiveRecord::Base.connection.reset_pk_sequence!('transactions')
    puts "Transactions loaded from CSV"
  end
```

**Review by ethanrossblack — CHANGES_REQUESTED** [review]
* Removed comments from everything except the first task
* Expanded the descriptions
* Removed the `.to_h` from when we create a new model. It still works as just `Customer.create(row)`

File: lib/tasks/csv_load.rake [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    file = "db/data/transactions.csv"
+    Transaction.destroy_all
+    #parse CSV file
+    CSV.foreach(file, headers: true) do |row|
+      #convert data into models for db
+      Transaction.create(row)
+    end
+    #reset pk sequence 
+    ActiveRecord::Base.connection.reset_pk_sequence!('transactions')
+    puts "Transactions loaded from CSV"
```
Comment (z-fitch) [inline lib/tasks/csv_load.rake:72]: I like the terminal responses for the rake tasks B)

File: lib/tasks/csv_load.rake [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+  end
+
+  desc "load all"
+    task all: do [:merchants, :customers, :items, :invoices, :invoice_items, :transactions]
```
Comment (z-fitch) [inline lib/tasks/csv_load.rake]: Should we add one here too? (terminal response )

**Review by z-fitch — APPROVED** [review]

**Review by z-fitch — APPROVED** [review]

**Review by ethanrossblack — APPROVED** [review]
Looks good now, approved!

**Review by JorjaF — COMMENTED** [review]
I looked over what you did and this looks really good

**Pull request merged**